### DISCUSSION TITLE
How to route user to correct app flow once user has selected a flow previously?

### DISCUSSION THREAD

**Discussion by samsnori** [react-navigation/react-navigation#11916] [Q&A]
This might be a bit off topic, but it's related to routing/navigation in general so please bear with me.

Assume you have an app where the user needs to login. After logging in the user is presented a couple of screens where he/she can select some options. Based on the selection we open a different section/screen in the app. 

Now the user closes the app and then reopens it again. As the user has been logged in already I would like to automatically load the correct section. How would you do this? I'm curious in both terms like where/how do you store the selection and how do you navigate? E.g. do you create a context which exposes methods that are used to cache the selection. Then when opening the app you check this context if there is any cache?

As an example, see this flow blow. The purplish screens are screens where the user can select his/her interests after which we load a different section/flow. 

<img width="1141" alt="Untitled" src="https://github.com/react-navigation/react-navigation/assets/156928941/6851432c-795a-46bc-9a6e-64fae43967d9">

**Comment by satya164** [comment]
You can provide a `initialState` prop to control which screens are shown on initial render https://reactnavigation.org/docs/7.x/navigation-container/#initialstate

**Reply by samsnori** [reply]
Thanks @satya164. 

When using `initialState` or `initialRouteName` it means that you can only mount your  navigation with components once you know what route to take, right?  I've tested this approach which works, however it also adds a delay/interruption when switching from the splash screen to the actual app flow. The reason for this, is because when you only mount all your components/navigations after the route is known the screens need to be created. 

Another approach I'm using now seems to prevent that. I have something like this:

```jsx
<AuthProvider>
  <Navigation />
  {auth_is_loading && 
    <SplashScreen />
 }
</AuthProvider>
```

With the above approach the navigation and its screen(s) are loaded and the splash screen is rendered at the top. 

To select the initial route, I created a "dummy" screen where I select the correct app flow. Imagine the following screen being the `initialRouteName` for the authenticated routes:

```jsx
function MyEntryScreen() {
   const auth_ctx = useAuthContext();
   
   let flow = auth_ctx.getFlowForLoggedInUser;
   switch (flow) {
     "flow11": { return (<Flow11Screen />); }
     "flow12": { return (<Flow12Screen />); }
     default: { return (<FlowSelection />); }
    }
}
```

**Reply by satya164** [reply]
I'm confused, how can you render the navigator if you don't even know which screen to render yet?

There are 2 cases, either you know what screen to render before auth and render the navigator with the `initialState` hidden behind the splash screen, or you need to auth first to know the screen, s you wait for auth before rendering the navigator.

**Reply by samsnori** [reply]
It would only be an issue when the navigator would render something, which I don't want the user to see. I can render the navigator because the splash screen is rendered on top of the navigator. The navigator will be hidden for the user until auth is ready. The `initialRouteName` will be used while the auth context is loading, however it doesn't really matter because the user doesn't see anything. 

For example when the user hasn't been logged in via the credentials which were saved (and loaded asynchronously) the navigator would render a login screen. But because this screen is behind the splash screen it's fine. Only after the cached credentials and some other state has been loaded I hide the splash screen.

I'm doing the first version of what you're saying: I know what screen to render before auth and render the navigator with `initialState`/`initialRouteName` hidden behind splash screen.  Because the navigator is hidden it doesn't matter what it's rendering. The splash screen is not part of the navigator.  When everything has been loaded, the `MyEntryScreen` will show the correct screen (which is the `initialEntryScreen`).
